### PR TITLE
Add database update system

### PR DESCRIPTION
**Pull request by following5**
This is a proposal for a database update system, to solve #1387. It's 2500 new lines of code and documentation, which distribute like this:

* 600 lines for the "DB update engine", including methods for synchronizing translations from files to DB tables
* 700 lines for tests and initial updates, which ensure identical DB structure on all OC sites
* 1200 lines for tools and methods to help developers writing and managing DB updates.

This all went through many iterations and tests, so I am pretty confident that it's stable and the code structure is basically ok.

Developer docs and workflow proposal is in [doc/DbUpdate.md](https://github.com/following5/opencaching-pl/blob/dbv2/docs/DbUpdate.md).

Deployment would be like this:

* Solve issue #1792 (see PR #1809).
* Merge this PR.
* Manually run the [001_test_OcDb](https://github.com/following5/opencaching-pl/blob/dbv2/Utils/Database/Updates/001_test_OcDb.php) (+ rollback) and [002_test_10_minutes](https://github.com/following5/opencaching-pl/blob/dbv2/Utils/Database/Updates/002_test_10_minutes.php) "updates" on all four OC sites (through the Admin.DbUpdate page). Both tests should not produce error messages and say "Ok".
* Add `util.sec/dbupdate.php` to the deployment scripts of all OC sites. The initial update [100_init](https://github.com/following5/opencaching-pl/blob/dbv2/Utils/Database/Updates/100_init.php) will run with the next code deployment.

If there are concerns regarding runtime restriction of the Github webhook that triggers code deployment, another test may be added for that and run before enabling `100_init`.

### CODE REVIEW

**Comment by following5** [comment]
Note on the files
* Config/user.default.php
* Config/user.uk.php
* lib/Objects/OcConfig/OcConfig.php
* lib/Objects/User/User.php

Opencache.UK has a special default setting for the `user.statpic_text` database field. To enable identical DB structure on all OC sites, this would be moved to a new config setting.

Columns which would be *deleted* by [100_init](https://github.com/following5/opencaching-pl/blob/dbv2/Utils/Database/Updates/100_init.php) (I could not find any reference to them in the source code):
* `notify_waiting.type` @ NL
* `sys_logins.success` @ PL, UK, RO
* `user.country`, `.get_bulletin`, `.hide_flag`, `.new_pw_date` @ NL

**Comment by following5** [comment]
> Opencache.UK has a special default setting for the `user.statpic_text database` field

This default is "Opencache.UK", compared to "Opencaching" for all other sites. So alternatively to introducing a setting, the site name could be used as default for all OC sites.

**Comment by kojoty** [comment]
@following5,
my feeling at first sight: wow :)

I assume that step:
"Manually run the 001_test_OcDb and 002_test_10_minutes "updates" on all four OC sites (through the Admin.DbUpdate page). Both tests should not produce error messages and say "Ok"." should be done at every dev. env. as well?

**Comment by following5** [comment]
Have rebased this on the new errorhandler (#1809), so that the full errorhandling can be tested here.

@kojoty 
> I assume that step:
> "Manually run the 001_test_OcDb and 002_test_10_minutes "updates" on all four OC sites (through the Admin.DbUpdate page). Both tests should not produce error messages and say "Ok"." should be done at every dev. env. as well?

Now that you mention it - yes. That should be done *now*, to test this change before merge.

Also, after 100_init was run at OC.PL and developer sites, developers should check if their databases are up to date or need more adjustments:
* http://local.opencaching.pl/okapi/devel/dbstruct?compare_to=https://opencaching.pl/okapi/devel/dbstruct

File: Utils/Database/Updates/UpdateScript.php [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+                !$this->db->columnExists($table, 'list_default_'.$lang)
+            ) {
+                $this->db->addColumnIfNotExists($table, 'list_default_'.$lang, 'int(1) NOT NULL DEFAULT 0');
+                $this->db->simpleQuery(
+                    "UPDATE `".$table."` SET `list_default_".$lang."` = list_default_pl"
+                );
+            }
+
+            if ($this->db->columnExists($table, 'sort_pl')) {
+                $this->db->addColumnIfNotExists($table, 'sort_'.$lang, 'varchar(128) NOT NULL NOT NULL');
```
Comment (following5) [inline Utils/Database/Updates/UpdateScript.php]: This is buggy - would reset all the `sort` colums in `countries` table. But at least parts of the current `sort` column contents also are broken ...

Comment (following5) [reply Utils/Database/Updates/UpdateScript.php]: Hrmpf.. did not understand my own code. :-o Everything ok here, will not replace anything.

**Comment by kojoty** [comment]
@deg-pl , @harrieklomp , @andrixnet 
Are you aware and ready for this change?!
Any opinion? Let's go with this!

**Comment by following5** [comment]
rebase on master / resolved merge conflict in an errorhandler comment ...

**Comment by kojoty** [comment]
@deg-pl @harrieklomp @andrixnet 
we are still waiting for your decision / opinion...

**Comment by harrieklomp** [comment]
No problem for me. I assume it is all okay :-)

**Comment by deg-pl** [comment]
I think it is good change. Let's merge.

**Comment by kojoty** [comment]
@harrieklomp, do you know what is going on with @andrixnet?
@following5 - do you think we can safety merge this PR if OCRO is temporary not cooperate?

**Comment by harrieklomp** [comment]
@kojoty 
Sometimes i do contact @andrixnet and he told me that he is very busy with his work and family. He did tell me that he is doing his best to read the latest changes and to cache up.

**Comment by following5** [comment]
> do you think we can safety merge this PR if OCRO doesn't cooperate temporarily?

Yes - merging, testing and installing at the three sites is safe. Also we can use it for some issues without OCRO cooperation, e.g. for the utf8mb4 migration. That can run later at RO site.

**Comment by kojoty** [comment]
> > do you think we can safety merge this PR if OCRO doesn't cooperate temporarily?
> 
> Yes - merging, testing and installing at the three sites is safe. Also we can use it for some issues without OCRO cooperation, e.g. for the utf8mb4 migration. That can run later at RO site.

hmm.. What I'm afraid of is that OCRO fetch all code changes and then it will also fetch db updates and I wonder if it is not a good idea to disable webhooks for OCRO - then we leave OCRO in stable state - when Andrei finds a time for maintenance of OCRO he just update its code...

What do you think about it?

(I would like to merge it after the weekend - it is an important change and I would like to not to make you for the next merge - and we have in queue the next quite big changes.

**Comment by following5** [comment]
I think the safest and most stable way is:

1. Merge this
2. run the two tests at PL, UK, NL sites
3. (fix it if there are problems)
4. install it in the deployment script at PL
5. test the automatic DB updating at PL
6. (fix it if there are problems)
7. install it in the deployment scripts at UK and NL sites
8. test automatic updating at UK and NL
9. When everything works at PL, UK and NL: **disable OCRO webhook**.

OCRO webhook will stay disabled until new cronjob engine and db updater are installed at OCRO.

**Comment by kojoty** [comment]
> I think the safest and most stable way is:
> 
> 1. Merge this
> 2. run the two tests at PL, UK, NL sites
> 3. (fix it if there are problems)
> 4. install it in the deployment script at PL
> 5. test the automatic DB updating at PL
> 6. (fix it if there are problems)
> 7. install it in the deployment scripts at UK and NL sites
> 8. test automatic updating at UK and NL
> 9. When everything works at PL, UK and NL: **disable OCRO webhook**.
> 
> OCRO webhook will stay disabled until new cronjob engine and db updater are installed at OCRO.

OK, let's go this way after the weekend :)
Thanks @following5 !

**Comment by kojoty** [comment]
> > I think the safest and most stable way is:
> > 
> > 1. Merge this
> > 2. run the two tests at PL, UK, NL sites
> > 3. (fix it if there are problems)
> > 4. install it in the deployment script at PL
> > 5. test the automatic DB updating at PL
> > 6. (fix it if there are problems)
> > 7. install it in the deployment scripts at UK and NL sites
> > 8. test automatic updating at UK and NL
> > 9. When everything works at PL, UK and NL: **disable OCRO webhook**.
> > 
> > OCRO webhook will stay disabled until new cronjob engine and db updater are installed at OCRO.
> 
> OK, let's go this way after the weekend :)

OK, change is merged
@following5 can you order directly what we should do now to pass step 2.

**Comment by following5** [comment]
* Go to https://opencaching.pl/Admin.DbUpdate
* [run] 001_test_OcDb -- should say "Ok" after some seconds
* [run] 002_test_10_minutes - should say "Ok" after ten minutes

**Comment by following5** [comment]
Then do the same for UK and NL. Your browser will block access to the website as long as 002_test_10_minutes is running (but you  may use another browser :-)

**Pull request merged**